sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -111,10 +112,15 @@ case class InMemoryTableScanExec(
 
   override def output: Seq[Attribute] = attributes
 
+  private def cachedPlan = relation.cachedPlan match {
+    case adaptive: AdaptiveSparkPlanExec if adaptive.isFinalized => adaptive.executedPlan
+    case other => other
```

Comment:
So this only affects output partitioning and ordering after the cached relation is materialized? And if the query plan refers this cached plan is already finished with planning but not executed yet, it will still use old partitioning and ordering, right?

Reply:
I think for good side, this can improve AQE coverage for some limited cases. Although I'm also worrying about some inconsistent behaviors that could be seen by end-users regarding shuffling/sorting. They might ask for questions why sometimes shuffle/sort is added but sometimes it isn't. 

I guess that might require users more tricks on optimizing cached AQE relation in practice.

Reply:
Another idea is to materialize the AQE plan eagerly so that even the first cache access can be optimized. However, this requires triggering query execution during query planning, which is a bit risky.

A good practice is to ask users to do query caching eagerly, e.g. do a `df.count` right after the `df` is cached. Then they won't observe inconsistencies. Anyway, I think this PR is a net win as it optimizes all the following cache accesses after the first access. This is important for query caching as cache is meant to be accessed repeatedly.

Reply:
Yes, this requires some good practice when caching query and using it. Otherwise, this looks a good one.